### ISSUE TITLE
Add support for German language

### ISSUE THREAD

**Issue by jwage** [doctrine/inflector#99]

**Comment by sinogermany** [comment]
That would be great if German is supported. It doesn't seem to be as straightforward though. Example:

- Singular: `Internationalisierungsstrategie`
- Plural: `Internationalisierungsstrategien`

How can we tell they are `Internationalisierung` and `Strategie` (Note there's a connecting `s`)?

Say we can - so would it become something like:

- camelCase (singular): `internationalisierungsStrategie` ???
- snake_case (singular): `internationalisierungs_strategie` ???

Never programmed in German, quite interested to see how it works. Let me know if I can be of any help.

**Comment by sinogermany** [comment]
A few compound word examples here:

- `Internationalisierungsstrategie` -> `/.+ung[s]?.+/`
- `Zwischenzeit`
- `Höchsttemperatur`
- `Haustürschlüsselloch`

Normally when we program in German how do we deal with umlauts? Use `e` instead?

**Comment by alcaeus** [comment]
> How can we tell they are `Internationalisierung` and `Strategie` (Note there's a connecting `s`)?

I may be missing something, but I don't think we need to match all words. For regular words, we only care about the ending. For irregular ones, we match from the end and keep building an ever longer and longer list of irregular words as issues appear.

> Normally when we program in German how do we deal with umlauts? Use `e` instead?

When people can't type them (e.g. because of restrictions in what characters are allowed), they normally use these replacements:
* `ä` => `ae`
* `ö` => `oe`
* `ü` => `ue`
* `ß` => `ss`

Some tools replace them with just the first letter of these replacements, which I believe is incorrect.

**Comment by jwage** [comment]
So who is going to add support for German? :) You can see the rules for another language here https://github.com/doctrine/inflector/tree/master/lib/Doctrine/Inflector/Rules/English

**Comment by dereuromark** [comment]
I would close this as impossible.
I once tried as well, and after 500 lines of exceptions and still no where close to having a reliable package this is just not worth doing. Whats the purpose if still every 2nd word doesn't work properly?
Inflection is not relevant here IMO, at least we have it well working for English.

**Comment by jwage** [comment]
Do you think it is really impossible? or just really hard with many exceptions to the rules? I would be interested still in getting a PR started and maybe we can slowly work on it over time and get contributions from people.

**Comment by dereuromark** [comment]
Knowing my German language well and looking into the issue it is really literally impossible.
It starts with words, that are the same but also need to know the "der/die/das" word in front of it to clarify the male/female/neutral form and thus the meaning, and as such also different plural forms for those. And that only covers this aspect, plurals on many many forms also require this male/female/neutral pre-word to have any possibility of building rules, and that is impossible if you only have the noun. So yeah it is a mess.

**Comment by jwage** [comment]
And if we had a public API that let you provide the word before, would it still be impossible in other cases?

**Comment by dereuromark** [comment]
You could start a library for all nouns. Together with this that would work.

**Comment by nschoellhorn** [comment]
Is this still relevant/wanted? I think the majority of words could be covered with a few rules. Of course, German is not an easy language so there will be quite some exceptions but they are there in English as well so I wouldn't say it's impossible. I can try to implement some of that, as far as I can get and then we see if it looks good or if there are any bigger hurdles to overcome.

**Comment by jwage** [comment]
I would still be interested in seeing support for other languages.

**Comment by nschoellhorn** [comment]
Ha, I indeed missed something, yes. Without the context of at least one sentence, it is not really doable without building the mentioned list of words. If we have a complete sentence, I guess it would look better, but with the word alone, we will not get very far. Sorry for digging this up needlessly :-/

**Comment by dereuromark** [comment]
Yeah, without the context (at least the "der/die/das" article) there will be no chance to translate properly.
Imagine "Leiter".

die Leiter ("ladder") => die Leitern
der Leiter ("leader") => die Leiter

And that is only one of a few issues to mention.

**Comment by jwage** [comment]
The API could be enhanced so that the context around the word can be passed through, no?

**Comment by dereuromark** [comment]
> Never programmed in German, quite interested to see how it works. Let me know if I can be of any help.

You also should NEVER program in German.^^

I only see the value here in translation tooling and maybe custom routing configs and other things that have context data processing, this should never find its way into actual PHP source code class files IMO.
And especially not as class names, method names, or other tokens.

**Comment by nschoellhorn** [comment]
> The API could be enhanced so that the context around the word can be passed through, no?

Yes, sure. But the problem is: how much of the context do you want to provide? There are cases where even the full sentence around the word might not be enough. I think it is doable if we get provided with the article of the given word. At least for the most cases. Would people want to provide that?

**Comment by grafst** [comment]
It is funny that this is almost trivial in english, but impossible in German.
